core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/JarResultBuildStep.java

```diff
@@ -1641,4 +1659,120 @@ public boolean decompile(Path jarToDecompile) {
         }
     }
 
+    public interface BytecodeOptimizer {
+        void init(Context context);
+
+        /**
+         * @return {@code true} if the optimizer was successfully download or already exists
+         */
+        boolean downloadIfNecessary();
+
+        /**
+         * @return {@code true} if the optimization process was successful
+         */
+        boolean optimize(Path jarToOptimize);
+
+        class Context {
+            final String versionStr;
+            final Path jarLocation;
+
+            public Context(String versionStr, Path jarLocation) {
+                this.versionStr = versionStr;
+                this.jarLocation = jarLocation;
+            }
+
+        }
+
+        class ProguardBytecodeOptimizer implements BytecodeOptimizer {
+            private Context context;
+            private Path proguardJar;
+
+            @Override
+            public void init(Context context) {
+                this.context = context;
+                this.proguardJar = context.jarLocation.resolve(String.format("proguard-%s.jar", context.versionStr));
+            }
+
+            @Override
+            public boolean downloadIfNecessary() {
+                if (Files.exists(proguardJar)) {
+                    return true;
+                }
+                String downloadURL = String.format(
+                        "https://github.com/Guardsquare/proguard/releases/download/v%s/proguard-%s.zip",
+                        context.versionStr, context.versionStr);
```

Comment:
This might require internet connection and access to Github everytime the build is run and ProGuard enabled (which is the default), I am thinking maybe we could make this an opt.
By "this" I mean, the proguard jar download. Allowing users to specify a path to their locally proguard if they've any. If not download.

Reply:
It has all kind of proxy limitations too. 

Is there really not maven coordinates for this artifact ?

Reply:
Yeah, there are. If we decide this is something we want, I'll update the PR to use Maven Central.